app/services/bulk_segmented_users.rb

```diff
@@ -0,0 +1,51 @@
+class BulkSegmentedUsers
+  def self.upsert(audience_segment, user_ids:)
+    return unless audience_segment.manual? && audience_segment.persisted?
+
+    now = Time.current
+    valid_user_ids = User.where(id: user_ids).ids
+
+    if valid_user_ids.empty?
+      return { succeeded: [], failed: user_ids }
+    end
+
+    segmented_users = valid_user_ids.map do |user_id|
+      {
+        audience_segment_id: audience_segment.id,
+        user_id: user_id,
+        created_at: now,
+        updated_at: now
+      }
+    end
+
+    upserted_user_ids = SegmentedUser.upsert_all(
+      segmented_users,
+      unique_by: :index_segmented_users_on_audience_segment_and_user,
+      update_only: [:updated_at],
+      record_timestamps: false,
```

Comment:
Minor: I found this surprising — seeing as we're setting `created_at` and `updated_at` explicitly on 16/17, I assume there's a reason we're turning `timestamps` off here?

Reply:
Ah - there is documentation for this class (especially for this decision) that didn't make it into this branch. I'll rectify that